### PR TITLE
server/api: add history api

### PR DESCRIPTION
**Pull request by nolouch**
PTAL @disksing @siddontang

### CODE REVIEW

File: server/api/history.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+	ops, err := h.GetHistoryOperatorsOfKind(server.ResourceKind(kind))
+	if err != nil {
+		h.r.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	if limit > len(ops) {
+		limit = len(ops)
+	}
+	h.r.JSON(w, http.StatusOK, ops[:limit])
```
Comment (disksing) [inline server/api/history.go]: Make sure _limit > 0_.

File: server/api/history_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	// gets history by kind and limit
+	tbl := []struct {
+		kind   int
+		limit  int
+		result int
+	}{
+		{0, 0, 0},
+		{0, 1, 1},
+		{0, 2, 2},
+		{0, 3, 2},
```
Comment (disksing) [inline server/api/history_test.go]: Need to check query another kind?

File: server/api/history.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		return
+	}
+
+	h.r.JSON(w, http.StatusOK, ops)
+}
+
+func (h *historyHandler) GetOperatorsOfKind(w http.ResponseWriter, r *http.Request) {
+	k := mux.Vars(r)["kind"]
+	l := mux.Vars(r)["limit"]
+	kind, err := strconv.Atoi(k)
```
Comment (disksing) [inline server/api/history.go]: Maybe it will be better to use strings for _kind_.

File: server/coordinator.go [hunk 4]
```diff
@@ -234,6 +234,7 @@ func (c *coordinator) addOperator(op Operator) bool {
 		log.Infof("coordinator: add operator %+v with higher priority, remove operator: %+v", op, old)
 	}
 
+	c.histories.add(regionID, op)
```
Comment (disksing) [inline server/coordinator.go]: The _op_ is added to history when _removeOperator_ too. https://github.com/pingcap/pd/pull/637/files#diff-414579b44396bf49088b1219e107ea44R262

File: server/api/history.go (hunk 3 above)
Comment (siddontang) [inline server/api/history.go]: +1

File: server/api/history.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	if err != nil {
+		h.r.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	limit, err := strconv.Atoi(l)
```
Comment (siddontang) [inline server/api/history.go]: can `limit` be 0 here? If yes, use a default value here.

File: server/store.go [hunk 6]
```diff
@@ -29,6 +29,7 @@ const (
 	leaderKind
 	regionKind
 	priorityKind
+	otherKind
```
Comment (siddontang) [inline server/store.go]: what does other mean?

File: server/coordinator.go [hunk 7]
```diff
@@ -291,6 +292,21 @@ func (c *coordinator) getHistories() []Operator {
...
 
+func (c *coordinator) getHistoriesOfKind(kind ResourceKind) []Operator {
+	c.RLock()
+	defer c.RUnlock()
+
+	var operators []Operator
+	for _, elem := range c.histories.elems() {
+		op := elem.value.(Operator)
+		if op.GetResourceKind() == kind {
+			operators = append(operators, elem.value.(Operator))
```
Comment (siddontang) [inline server/coordinator.go]: /elem.value.(Operator)/op/

File: server/coordinator.go (hunk 4 above)
Comment (nolouch) [inline server/coordinator.go]: I need to know whether the operator running, finish or timeout. so we need to add it in here.

File: server/store.go (hunk 6 above)
Comment (nolouch) [inline server/store.go]: now the split record as a operator, I define it as other kind

File: server/api/history_test.go (hunk 2 above)
Comment (nolouch) [inline server/api/history_test.go]: I think the test logic is already covered, and now we just have an exposed interface to add an admin kind operator

File: server/api/history.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	h.r.JSON(w, http.StatusOK, ops)
+}
+
+func (h *historyHandler) GetOperatorsOfKind(w http.ResponseWriter, r *http.Request) {
+	k := mux.Vars(r)["kind"]
+	l := mux.Vars(r)["limit"]
+	var kind server.ResourceKind
+	switch k {
```
Comment (siddontang) [inline server/api/history.go]: add a NameToKind function

Btw, I suggest using kind 0 for Unknown, not for admin.

**Comment by siddontang** [comment]
any update? @nolouch

File: server/operator.go [hunk 9]
```diff
@@ -22,10 +22,61 @@ import (
...
+var resourceKindToName = map[int]string{
+	0: "unknown",
+	1: "admin",
+	2: "leader",
+	3: "region",
+	4: "priority",
+	5: "other",
+}
+
+var resourceKindToValue = map[string]ResourceKind{
```
Comment (siddontang) [inline server/operator.go]: resourceNameToKind

File: server/operator.go [hunk 10]
```diff
@@ -22,10 +22,61 @@ import (
...
+	return resourceKindToName[0]
+}
+
+// ParseResourceKind convert string to ResourceKind
+func ParseResourceKind(name string) ResourceKind {
+	k, ok := resourceKindToValue[name]
+	if ok {
+		return k
+	}
+	return resourceKindToValue["unknown"]
```
Comment (siddontang) [inline server/operator.go]: return UnKonwKind here directly

File: server/operator.go [hunk 11]
```diff
@@ -22,10 +22,61 @@ import (
...
+var resourceKindToValue = map[string]ResourceKind{
+	"unknown":  UnKnownKind,
+	"admin":    AdminKind,
+	"leader":   LeaderKind,
+	"region":   RegionKind,
+	"priority": PriorityKind,
+	"other":    OtherKind,
+}
+
+func (k ResourceKind) String() string {
```
Comment (siddontang) [inline server/operator.go]: please add test for it

**Comment by siddontang** [comment]
PTAL @disksing

File: server/operator_test.go [hunk 12]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (disksing) [inline server/operator_test.go]: Now it's 2017.

**Pull request merged**